build/provenance.go

```diff
@@ -0,0 +1,198 @@
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/containerd/containerd/content/proxy"
+	"github.com/docker/buildx/util/confutil"
+	"github.com/docker/buildx/util/progress"
+	controlapi "github.com/moby/buildkit/api/services/control"
+	"github.com/moby/buildkit/client"
+	provenancetypes "github.com/moby/buildkit/solver/llbsolver/provenance/types"
+	"github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+func setRecordProvenance(ctx context.Context, c *client.Client, sr *client.SolveResponse, ref string, pw progress.Writer) error {
+	mode := confutil.MetadataProvenance()
+	if mode == confutil.MetadataProvenanceModeNone {
+		return nil
+	}
+	pw = progress.ResetTime(pw)
+	return progress.Wrap("resolve build record provenance", pw.Write, func(l progress.SubLogger) error {
+		bo := backoff.NewExponentialBackOff()
+		bo.MaxElapsedTime = 5 * time.Second
+		return backoff.Retry(func() error {
+			res, err := fetchProvenance(ctx, c, ref, mode, l)
+			if err != nil {
+				return err
+			}
+			for k, v := range res {
+				sr.ExporterResponse[k] = v
+			}
+			return nil
+		}, bo)
+	})
+}
+
+func fetchProvenance(ctx context.Context, c *client.Client, ref string, mode confutil.MetadataProvenanceMode, l progress.SubLogger) (out map[string]string, err error) {
+	cl, err := c.ControlClient().ListenBuildHistory(ctx, &controlapi.BuildHistoryRequest{
+		Ref:       ref,
+		EarlyExit: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var mu sync.Mutex
+	eg, ctx := errgroup.WithContext(ctx)
+	for {
+		ev, err := cl.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		if ev.Record == nil {
+			continue
+		}
+		if ev.Record.Result != nil {
+			provenanceDgst := provenanceDigest(ev.Record.Result)
+			if provenanceDgst == nil {
+				continue
+			}
+			eg.Go(func() error {
+				dt, err := getBlob(ctx, c, *provenanceDgst, "", l)
+				if err != nil {
+					return errors.Wrapf(err, "failed to load provenance from build record")
+				}
+				prv, err := encodeProvenance(dt, mode)
+				if err != nil {
+					return err
+				}
+				mu.Lock()
+				if out == nil {
+					out = make(map[string]string)
+				}
+				out["buildx.build.provenance"] = prv
+				mu.Unlock()
+				return nil
+			})
+		} else if ev.Record.Results != nil {
+			for platform, res := range ev.Record.Results {
+				platform := platform
+				provenanceDgst := provenanceDigest(res)
+				if provenanceDgst == nil {
+					continue
+				}
+				eg.Go(func() error {
+					dt, err := getBlob(ctx, c, *provenanceDgst, platform, l)
+					if err != nil {
+						return errors.Wrapf(err, "failed to load provenance from build record")
+					}
+					prv, err := encodeProvenance(dt, mode)
+					if err != nil {
+						return err
+					}
+					mu.Lock()
+					if out == nil {
+						out = make(map[string]string)
+					}
+					out["buildx.build.provenance/"+platform] = prv
```

Comment:
This doesn't seem to work. I don't get provenance keys for multi-platform builds when testing.

One thing that is bit weird in here as well is that the metadata-file doesn't currently contain the image digest for a specific platform in multi-platform builds, so digest for the artifact that the provenance is for.

Reply:
> This doesn't seem to work. I don't get provenance keys for multi-platform builds when testing.

What cmd did you invoke? It works for me (see second example in PR desc) with:

```
docker buildx --builder builder bake binaries-cross --metadata-file md.json
```

```json
{
  "binaries-cross": {
    "buildx.build.provenance/darwin/amd64": {
      "builder": {
        "id": ""
      },
      "buildType": "https://mobyproject.org/buildkit@v1",
      "materials": [
        {
          "uri": "pkg:docker/docker/dockerfile@1",
          "digest": {
            "sha256": "dbbd5e059e8a07ff7ea6233b213b36aa516b4c53c645f1817a4dd18b83cbea56"
          }
        },
        {
          "uri": "pkg:docker/golang@1.21-alpine?platform=linux%2Famd64",
          "digest": {
            "sha256": "ed8ce6c22dd111631c062218989d17ab4b46b503cbe9a9cfce1517836e65298a"
          }
        },
        {
          "uri": "pkg:docker/tonistiigi/xx@1.4.0?platform=linux%2Famd64",
          "digest": {
            "sha256": "0cd3f05c72d6c9b038eb135f91376ee1169ef3a330d34e418e65e2a5c2e9c0d4"
          }
        }
      ],
      "invocation": {
        "configSource": {},
        "parameters": {
          "frontend": "gateway.v0",
          "args": {
            "build-arg:BUILDKIT_CONTEXT_KEEP_GIT_DIR": "1",
            "cmdline": "docker/dockerfile:1",
            "source": "docker/dockerfile:1",
            "target": "binaries"
          },
          "locals": [
            {
              "name": "context"
            },
            {
              "name": "dockerfile"
            }
          ]
        },
        "environment": {
          "platform": "linux/amd64"
        }
      }
    },
    "buildx.build.provenance/darwin/arm64": {
      "builder": {
        "id": ""
      },
      "buildType": "https://mobyproject.org/buildkit@v1",
      "materials": [
        {
          "uri": "pkg:docker/docker/dockerfile@1",
          "digest": {
            "sha256": "dbbd5e059e8a07ff7ea6233b213b36aa516b4c53c645f1817a4dd18b83cbea56"
          }
        },
        {
          "uri": "pkg:docker/golang@1.21-alpine?platform=linux%2Famd64",
          "digest": {
            "sha256": "ed8ce6c22dd111631c062218989d17ab4b46b503cbe9a9cfce1517836e65298a"
          }
        },
        {
          "uri": "pkg:docker/tonistiigi/xx@1.4.0?platform=linux%2Famd64",
          "digest": {
            "sha256": "0cd3f05c72d6c9b038eb135f91376ee1169ef3a330d34e418e65e2a5c2e9c0d4"
          }
        }
      ],
      "invocation": {
        "configSource": {},
        "parameters": {
          "frontend": "gateway.v0",
          "args": {
            "build-arg:BUILDKIT_CONTEXT_KEEP_GIT_DIR": "1",
            "cmdline": "docker/dockerfile:1",
            "source": "docker/dockerfile:1",
            "target": "binaries"
          },
          "locals": [
            {
              "name": "context"
            },
            {
              "name": "dockerfile"
            }
          ]
        },
        "environment": {
          "platform": "linux/amd64"
        }
      }
    },
    "buildx.build.provenance/linux/amd64": {
      "builder": {
        "id": ""
      },
      "buildType": "https://mobyproject.org/buildkit@v1",
      ...
```

Reply:
> One thing that is bit weird in here as well is that the metadata-file doesn't currently contain the image digest for a specific platform in multi-platform builds, so digest for the artifact that the provenance is for.

Yes it doesn't seem we set them in exporter response, this is slightly related to https://github.com/moby/buildkit/issues/4532